File: scripts/lint_pack.py
"""Static linter for agent pack contracts.

Replaces the per-case L1/L2/L3 assertions from the legacy YAML harness.
This runs **once per pack** (not once per eval) and validates that each
``.agent.md`` file has well-formed YAML front-matter declaring the fields
the multi-agent system relies on.

Why a static linter rather than per-eval assertions?

* Pack contracts (allowed tools, scope boundaries, descriptions) are
  pack-level facts, not case-level. Validating them once is enough.
* Static linting catches violations at edit time without paying for an
  LLM run.
* Plain Python is easy to read and modify; tests don't have to encode
  regex grammars.

Usage::

    python -m scripts.lint_pack <pack>            # one pack
    python -m scripts.lint_pack --all             # every pack under agent-packs/

Exit codes: 0 = clean, 1 = violations found, 2 = bad invocation.
"""

from __future__ import annotations

import argparse
import dataclasses
import sys
from pathlib import Path
from typing import Iterable

import yaml


REPO_ROOT = Path(__file__).resolve().parents[1]
PACKS_ROOT = REPO_ROOT / "agent-packs"


# Required front-matter keys for every .agent.md file. These mirror the
# schema GitHub Copilot CLI itself enforces, plus a few sanity checks.
REQUIRED_AGENT_FIELDS = ("description", "tools")

# Tools allow-list. Any tool name outside this set is flagged.
KNOWN_TOOLS = {
    "read", "edit", "search", "execute", "agent", "shell",
    "create", "view", "grep", "glob", "task",
    "write", "delete", "list", "session_store_sql",
    "web_fetch", "web_search", "web", "data",
    "*",  # wildcard: agent opts into any tool
}

# Soft caps from the factory's quality standards. Warnings, not errors.
AGENT_MAX_CHARS = 30_000
SKILL_MAX_WORDS = 5_000


@dataclasses.dataclass
class Issue:
    severity: str  # "error" | "warn"
    path: Path
    message: str

    def format(self) -> str:
        rel = self.path.relative_to(REPO_ROOT) if self.path.is_absolute() else self.path
        return f"{self.severity.upper():5s} {rel}: {self.message}"


def is_agent_pack(pack_dir: Path) -> bool:
    """A pack is an agent pack iff it ships ``.github/agents/`` or
    ``.github/skills/``. Roo-only packs (``.roomodes`` / ``.roo/``) are
    out of scope for this linter.
    """
    return (
        (pack_dir / ".github" / "agents").exists()
        or (pack_dir / ".github" / "skills").exists()
    )


def lint_pack(pack_dir: Path) -> list[Issue]:
    """Return the list of issues found in ``pack_dir``.

    Non-Copilot packs (Roo-only) return an empty list -- they are out of
    scope for this linter, which validates Copilot CLI agent contracts.
    """
    issues: list[Issue] = []

    if not pack_dir.is_dir():
        return [Issue("error", pack_dir, "Pack directory does not exist")]

    if not is_agent_pack(pack_dir):
        # Roo-only or unknown layout; skip silently. The caller can use
        # ``is_agent_pack`` to decide whether to include the pack.
        return []

    if not (pack_dir / "README.md").exists():
        issues.append(Issue("error", pack_dir, "Pack is missing README.md"))

    agents_dir = pack_dir / ".github" / "agents"
    skills_dir = pack_dir / ".github" / "skills"

    if agents_dir.exists():
        for agent_file in sorted(agents_dir.glob("*.agent.md")):
            issues.extend(_lint_agent_file(agent_file))

    if skills_dir.exists():
        for skill_md in sorted(skills_dir.rglob("SKILL.md")):
            issues.extend(_lint_skill_file(skill_md))

    return issues


def _lint_agent_file(path: Path) -> list[Issue]:
    issues: list[Issue] = []
    text = path.read_text(encoding="utf-8", errors="replace")

    if len(text) > AGENT_MAX_CHARS:
        issues.append(
            Issue(
                "warn", path,
                f"Agent file is {len(text)} chars (>{AGENT_MAX_CHARS} soft cap)",
            )
        )

    front_matter, body = _split_front_matter(text)
    if front_matter is None:
        return issues + [Issue("error", path, "Missing YAML front-matter (--- block)")]

    try:
        meta = yaml.safe_load(front_matter) or {}
    except yaml.YAMLError as exc:
        return issues + [Issue("error", path, f"Invalid YAML front-matter: {exc}")]

    if not isinstance(meta, dict):
        return issues + [Issue("error", path, "Front-matter must be a mapping")]

    for field in REQUIRED_AGENT_FIELDS:
        if field not in meta:
            issues.append(Issue("error", path, f"Front-matter missing required key: {field}"))

    desc = meta.get("description", "")
    if isinstance(desc, str) and len(desc.strip()) < 30:
        issues.append(
            Issue(
                "warn", path,
                "Description is very short (< 30 chars); descriptions need triggers/keywords",
            )
        )

    tools = meta.get("tools")
    if tools is not None:
        if not isinstance(tools, list):
            issues.append(Issue("error", path, "tools must be a YAML list"))
        else:
            for t in tools:
                if not isinstance(t, str):
                    issues.append(Issue("error", path, f"tools entry not a string: {t!r}"))
                    continue
                # Allow MCP-prefixed tools (server-tool form).
                if t in KNOWN_TOOLS or "-" in t:
                    continue
                issues.append(Issue("warn", path, f"Unknown tool: {t!r}"))

    if not body.strip():
        issues.append(Issue("error", path, "Agent body is empty"))

    return issues


def _lint_skill_file(path: Path) -> list[Issue]:
    issues: list[Issue] = []
    text = path.read_text(encoding="utf-8", errors="replace")
    word_count = len(text.split())
    if word_count > SKILL_MAX_WORDS:
        issues.append(
            Issue(
                "warn", path,
                f"Skill is {word_count} words (>{SKILL_MAX_WORDS} soft cap); "
                "consider splitting into reference docs",
            )
        )
    return issues


def _split_front_matter(text: str) -> tuple[str | None, str]:
    """Return ``(front_matter_yaml, body)`` or ``(None, full_text)``."""
    if not text.startswith("---"):
        return None, text
    end = text.find("\n---", 3)
    if end == -1:
        return None, text
    return text[3:end].lstrip("\n"), text[end + 4 :].lstrip("\n")


def discover_packs() -> list[Path]:
    """Return Copilot packs only (skips Roo-only packs)."""
    if not PACKS_ROOT.exists():
        return []
    return sorted(
        p for p in PACKS_ROOT.iterdir() if p.is_dir() and is_agent_pack(p)
    )


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("pack", nargs="?", help="Name of pack to lint (under agent-packs/)")
    p.add_argument("--all", action="store_true", help="Lint every pack")
    p.add_argument(
        "--strict", action="store_true",
        help="Treat warnings as errors (exit 1 on any warn)",
    )
    args = p.parse_args(argv)

    if not args.all and not args.pack:
        p.error("Specify a pack name or --all")

    packs = discover_packs() if args.all else [PACKS_ROOT / args.pack]
    all_issues: list[Issue] = []
    for pack in packs:
        all_issues.extend(lint_pack(pack))

    for issue in all_issues:
        print(issue.format())

    has_errors = any(i.severity == "error" for i in all_issues)
    has_warns = any(i.severity == "warn" for i in all_issues)
    if has_errors or (args.strict and has_warns):
        print(f"\n{len(all_issues)} issue(s) found.", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/lint-pack.mjs
/**
 * Static linter for agent pack contracts (dependency-free Node port of the
 * legacy `scripts/lint_pack.py`).
 *
 * Runs once per pack (not once per eval) and validates that each `.agent.md`
 * file has well-formed YAML front-matter declaring the fields the multi-agent
 * system relies on. Consumed both from the CLI and from the structural eval
 * `evals/static/pack_contract.eval.ts`.
 *
 * Usage:
 *   node scripts/lint-pack.mjs <pack>     # one pack under agent-packs/
 *   node scripts/lint-pack.mjs --all      # every pack
 *
 * Exit codes: 0 = clean, 1 = violations found, 2 = bad invocation.
 */

import { readFileSync, existsSync, statSync, readdirSync } from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
export const REPO_ROOT = path.resolve(HERE, "..");
const PACKS_ROOT = path.join(REPO_ROOT, "agent-packs");

// Required front-matter keys for every .agent.md file.
const REQUIRED_AGENT_FIELDS = ["description", "tools"];

// Tools allow-list. Any tool name outside this set is flagged (warn).
const KNOWN_TOOLS = new Set([
  "read", "edit", "search", "execute", "agent", "shell",
  "create", "view", "grep", "glob", "task",
  "write", "delete", "list", "session_store_sql",
  "web_fetch", "web_search", "web", "data",
  "*",
]);

// Soft caps from the factory's quality standards (warnings, not errors).
const AGENT_MAX_CHARS = 30_000;
const SKILL_MAX_WORDS = 5_000;

export class Issue {
  /** @param {"error"|"warn"} severity @param {string} filePath @param {string} message */
  constructor(severity, filePath, message) {
    this.severity = severity;
    this.path = filePath;
    this.message = message;
  }

  format() {
    let rel = this.path;
    if (path.isAbsolute(rel)) rel = path.relative(REPO_ROOT, rel);
    rel = rel.split(path.sep).join("/");
    return `${this.severity.toUpperCase().padEnd(5)} ${rel}: ${this.message}`;
  }
}

/** A pack is an agent pack iff it ships `.github/agents/` or `.github/skills/`. */
export function isAgentPack(packDir) {
  return (
    existsSync(path.join(packDir, ".github", "agents")) ||
    existsSync(path.join(packDir, ".github", "skills"))
  );
}

function isDir(p) {
  try {
    return statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function globFiles(dir, predicate, recurse) {
  /** @type {string[]} */
  const out = [];
  if (!isDir(dir)) return out;
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (recurse) out.push(...globFiles(full, predicate, recurse));
    } else if (predicate(entry.name)) {
      out.push(full);
    }
  }
  return out.sort();
}

/** @returns {Issue[]} */
export function lintPack(packDir) {
  /** @type {Issue[]} */
  const issues = [];

  if (!isDir(packDir)) {
    return [new Issue("error", packDir, "Pack directory does not exist")];
  }
  if (!isAgentPack(packDir)) {
    return []; // Roo-only / unknown layout: out of scope, skip silently.
  }
  if (!existsSync(path.join(packDir, "README.md"))) {
    issues.push(new Issue("error", packDir, "Pack is missing README.md"));
  }

  const agentsDir = path.join(packDir, ".github", "agents");
  const skillsDir = path.join(packDir, ".github", "skills");

  for (const agentFile of globFiles(agentsDir, (n) => n.endsWith(".agent.md"), false)) {
    issues.push(...lintAgentFile(agentFile));
  }
  for (const skillMd of globFiles(skillsDir, (n) => n === "SKILL.md", true)) {
    issues.push(...lintSkillFile(skillMd));
  }
  return issues;
}

function lintAgentFile(filePath) {
  /** @type {Issue[]} */
  const issues = [];
  const text = readFileSync(filePath, "utf-8");

  if (text.length > AGENT_MAX_CHARS) {
    issues.push(
      new Issue(
        "warn",
        filePath,
        `Agent file is ${text.length} chars (>${AGENT_MAX_CHARS} soft cap)`,
      ),
    );
  }

  const [frontMatter, body] = splitFrontMatter(text);
  if (frontMatter === null) {
    issues.push(new Issue("error", filePath, "Missing YAML front-matter (--- block)"));
    return issues;
  }

  let meta;
  try {
    meta = parseFrontMatter(frontMatter);
  } catch (exc) {
    issues.push(new Issue("error", filePath, `Invalid YAML front-matter: ${exc}`));
    return issues;
  }
  if (meta === null || typeof meta !== "object" || Array.isArray(meta)) {
    issues.push(new Issue("error", filePath, "Front-matter must be a mapping"));
    return issues;
  }

  for (const field of REQUIRED_AGENT_FIELDS) {
    if (!(field in meta)) {
      issues.push(new Issue("error", filePath, `Front-matter missing required key: ${field}`));
    }
  }

  const desc = meta.description ?? "";
  if (typeof desc === "string" && desc.trim().length < 30) {
    issues.push(
      new Issue(
        "warn",
        filePath,
        "Description is very short (< 30 chars); descriptions need triggers/keywords",
      ),
    );
  }

  const tools = meta.tools;
  if (tools !== undefined && tools !== null) {
    if (!Array.isArray(tools)) {
      issues.push(new Issue("error", filePath, "tools must be a YAML list"));
    } else {
      for (const t of tools) {
        if (typeof t !== "string") {
          issues.push(new Issue("error", filePath, `tools entry not a string: ${JSON.stringify(t)}`));
          continue;
        }
        if (KNOWN_TOOLS.has(t) || t.includes("-")) continue; // allow MCP-prefixed tools
        issues.push(new Issue("warn", filePath, `Unknown tool: ${JSON.stringify(t)}`));
      }
    }
  }

  if (!body.trim()) {
    issues.push(new Issue("error", filePath, "Agent body is empty"));
  }
  return issues;
}

function lintSkillFile(filePath) {
  /** @type {Issue[]} */
  const issues = [];
  const text = readFileSync(filePath, "utf-8");
  const wordCount = text.split(/\s+/).filter(Boolean).length;
  if (wordCount > SKILL_MAX_WORDS) {
    issues.push(
      new Issue(
        "warn",
        filePath,
        `Skill is ${wordCount} words (>${SKILL_MAX_WORDS} soft cap); ` +
          "consider splitting into reference docs",
      ),
    );
  }
  return issues;
}

/** @returns {[string|null, string]} (frontMatterYaml, body) or (null, fullText). */
function splitFrontMatter(text) {
  if (!text.startsWith("---")) return [null, text];
  const end = text.indexOf("\n---", 3);
  if (end === -1) return [null, text];
  return [text.slice(3, end).replace(/^\n+/, ""), text.slice(end + 4).replace(/^\n+/, "")];
}

function stripQuotes(v) {
  const s = v.trim();
  if (s.length >= 2 && ((s[0] === '"' && s.endsWith('"')) || (s[0] === "'" && s.endsWith("'")))) {
    return s.slice(1, -1);
  }
  return s;
}

/**
 * Minimal YAML front-matter parser: flat `key: value` pairs, block lists
 * (`- item`), and flow lists (`[a, b]`). Sufficient for `.agent.md` schemas;
 * avoids a yaml dependency so the linter runs with zero installs.
 * @returns {Record<string, unknown>}
 */
export function parseFrontMatter(fm) {
  const lines = fm.replace(/\r\n?/g, "\n").split("\n");
  /** @type {Record<string, unknown>} */
  const out = {};
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (!line.trim() || line.trimStart().startsWith("#")) continue;
    if (/^\s/.test(line)) continue; // handled by look-ahead below
    const m = /^([^:\s][^:]*):(.*)$/.exec(line);
    if (!m) continue;
    const key = m[1].trim();
    const rest = m[2].trim();
    if (rest === "") {
      // Block list or nested mapping via look-ahead.
      /** @type {string[]} */
      const items = [];
      let isMapping = false;
      let j = i + 1;
      for (; j < lines.length; j++) {
        const nxt = lines[j];
        if (!nxt.trim()) continue;
        if (!/^\s/.test(nxt)) break; // dedent -> next top-level key
        const li = /^\s+-\s*(.*)$/.exec(nxt);
        if (li) {
          items.push(stripQuotes(li[1]));
        } else {
          isMapping = true; // indented `k: v` -> nested map (contents ignored)
        }
      }
      out[key] = isMapping && items.length === 0 ? {} : items;
      i = j - 1;
    } else if (rest.startsWith("[") && rest.endsWith("]")) {
      const inner = rest.slice(1, -1).trim();
      out[key] = inner ? inner.split(",").map((x) => stripQuotes(x)) : [];
    } else {
      const scalar = stripQuotes(rest);
      if (scalar === "true") out[key] = true;
      else if (scalar === "false") out[key] = false;
      else out[key] = scalar;
    }
  }
  return out;
}

/** Return Copilot packs only (skips Roo-only packs), sorted by path. */
export function discoverPacks() {
  if (!isDir(PACKS_ROOT)) return [];
  return readdirSync(PACKS_ROOT, { withFileTypes: true })
    .filter((e) => e.isDirectory())
    .map((e) => path.join(PACKS_ROOT, e.name))
    .filter(isAgentPack)
    .sort();
}

function main(argv) {
  const args = argv.slice(2);
  const all = args.includes("--all");
  const strict = args.includes("--strict");
  const positional = args.filter((a) => !a.startsWith("--"));

  if (!all && positional.length === 0) {
    console.error("Specify a pack name or --all");
    return 2;
  }

  const packs = all ? discoverPacks() : [path.join(PACKS_ROOT, positional[0])];
  /** @type {Issue[]} */
  const allIssues = [];
  for (const pack of packs) allIssues.push(...lintPack(pack));

  for (const issue of allIssues) console.log(issue.format());

  const hasErrors = allIssues.some((i) => i.severity === "error");
  const hasWarns = allIssues.some((i) => i.severity === "warn");
  if (hasErrors || (strict && hasWarns)) {
    console.error(`\n${allIssues.length} issue(s) found.`);
    return 1;
  }
  return 0;
}

// Run as a script (not when imported by an eval).
if (import.meta.url === `file://${process.argv[1]}` || fileURLToPath(import.meta.url) === process.argv[1]) {
  process.exit(main(process.argv));
}
